src/scheduler/MpiWorld.cpp

```diff
@@ -1571,7 +1577,31 @@ void MpiWorld::prepareMigration(
         for (int i = 0; i < pendingMigrations->migrations_size(); i++) {
             auto m = pendingMigrations->mutable_migrations()->at(i);
             assert(hostForRank.at(m.msg().mpirank()) == m.srchost());
+            // Update the host for this rank. We only update the positions of
+            // the to-be migrated ranks, avoiding race conditions with not-
+            // migrated ranks
             hostForRank.at(m.msg().mpirank()) = m.dsthost();
+            // Update the ranks for host. This structure is used when doing
+            // collective communications by all ranks. At this point, all non-
+            // leader ranks will be hitting a barrier, for which they don't
+            // need the ranks for host map, therefore it is safe to modify it
+            if (m.dsthost() == thisHost && m.msg().mpirank() < localLeader) {
```

Comment:
We used to call `initLocalRemote()` here, unfortunately that method is too coarse and was clearing the `rankForHost` structure before filling it again.

This was causing a race condition between migrated and non-migrated ranks.